File: sort/python_implement/test_sort.py
#!/usr/bin/env python3
# coding: utf-8
#
# author: kaku
# date: 19/10/07

# GitHub:
#
#   https://github.com/kakukosaku
#
# © 2019-2022 Kaku Kosaku All Rights Reserved
#
# Usage:
#   python -m python_implement.test_sort

from typing import List, NoReturn, Callable

from .array import random_array, show_array, check_sorted
from . import bubble_sort, insert_sort, select_sort

ARRAY_SIZE = 10


def test_case(arr: List[int], array_size: int, description: str, sort_func: Callable) -> NoReturn:
    _arr = list(arr)
    sort_func(_arr, array_size)
    show_array(_arr, description)
    check_sorted(_arr)


def main():
    arr = random_array(ARRAY_SIZE)
    show_array(arr, "Original Array")

    test_case(arr, ARRAY_SIZE, "After Bubble Sort", bubble_sort)
    test_case(arr, ARRAY_SIZE, "After Insert Sort", insert_sort)
    test_case(arr, ARRAY_SIZE, "After Select Sort", select_sort)


if __name__ == '__main__':
    main()

File: sort/c_impl/test_sort.c
//
// Created by kaku on 2019/10/7.
//

#include <stdlib.h>
#include "array.h"

#define ARRAY_SIZE 10

void test_case(int *arr, int array_size, const char description[], void (*func)(int *, int)) {
    int *temp_arr;
    temp_arr = copy_array(arr, ARRAY_SIZE);
    func(temp_arr, array_size);
    show_array(temp_arr, array_size, description);
    check_sorted(temp_arr, array_size, False);
    free(temp_arr);
}

int main() {
    int *arr;
    arr = random_array(ARRAY_SIZE);
    show_array(arr, ARRAY_SIZE, "Original Array");

    test_case(arr, ARRAY_SIZE, "After Bubble Sort", bubble_sort);
    test_case(arr, ARRAY_SIZE, "After Select Sort", select_sort);
    test_case(arr, ARRAY_SIZE, "After Insert Sort", insert_sort);
    test_case(arr, ARRAY_SIZE, "After Quick Sort", quick_sort);
    test_case(arr, ARRAY_SIZE, "After Heap Sort", heap_sort);
}

